flocker/acceptance/test_restarts.py

```diff
@@ -0,0 +1,109 @@
+
+from ..testtools import loop_until, random_name
+from .testtools import create_dataset, require_cluster
+
+
+REBOOT_SERVER = FilePath(__file__).sibling(b"reboot_httpserver.py")
+
+
+class RestartTests(TestCase):
+    """
+    Tests for restart policies.
+    """
+    @require_cluster(1)
+    def test_restart_always_reboot_with_dataset(self, cluster):
+        """
+        If a container has a restart policy of ``always`` and a volume mapped
+        to a dataset, on reboots it will only be restarted after the
+        volume becomes available.
+        """
+        if not os.environ.get("RUN_REBOOT_TESTS"):
+            raise SkipTest(
+                    "Don't want to run this on buildbot, for now at least.")
+        node = cluster.nodes[0]
+        # Implicitly uses first node:
+        creating_dataset = create_dataset(self, cluster)
+
+        def query_server():
+            req = get(
+                "http://{host}:12345".format(host=node.public_address),
+                persistent=False
+            ).addCallbacks(content)
+            return req
+
+        def created_dataset(dataset):
+            print "DATASET", dataset
+            http_server = {
+                u"name": random_name(self),
+                u"node_uuid": str(cluster.nodes[0].uuid),
+                u"image": u"python:2.7-slim",
+                u"ports": [{u"internal": 12345, u"external": 12345}],
+                # We expect restarts to occur after reboot with this policy:
+                u'restart_policy': {u'name': u'always'},
+                u"volumes": [{u"dataset_id": str(dataset.dataset_id),
+                              u"mountpoint": u"/data"}],
+                u"command_line": [u"python", u"-c",
+                                  REBOOT_SERVER.getContent().decode("ascii"),
+                                  u"/data"],
+            }
+            created = cluster.create_container(http_server)
+            created.addCallback(lambda _: self.addCleanup(
+                cluster.remove_container, http_server[u"name"]))
+            created.addCallback(lambda _:
+                                loop_until(lambda: query_server().addErrback(
+                                    lambda _: False)))
+            created.addCallback(lambda _: query_server())
+            return created
+        creating_dataset.addCallback(created_dataset)
+
+        def server_started(initial_response):
+            # We now have the initial response of the server. This should
+            # be the initial boot time of the node, repeated twice.  Next,
+            # we restart the machine.
+            print "INITIAL RESPONSE", initial_response
+            initial_reboot_time = initial_response.splitlines()[0]
+            # XXX Checking exit code is problematic insofar as reboot
+            # kills the ssh process...
+            print "Rebooting!"
+            call([b"ssh", b"root@{}".format(node.public_address),
+                  b"shutdown", b"-r", b"now"])
```

Comment:
Although, my experience is that calling `shutdown -r now` will kill the remote SSH process and cause the local SSH process to exit with an error status....which will cause `run_ssh` to raise an exception which we'll need to handle.